### PR TITLE
Prevent non-convex errors from appearing when using PixelCounting

### LINKED ISSUE #8262: Severe "non-convex" errors appear when using PixelCounting
Issue overview
--------------
The PixelCounting method tessellates all surfaces into many triangles (which are by definition convex), removing any limitations of non-convex (i.e., concave) surfaces from the shading algorithm. These errors should not appear.

See this post on [Unmet Hours](https://unmethours.com/question/47154/can-i-ignore-the-severe-non-convex-determineshadowingcombinations-warning-when-using-pixelcounting/).

Defect file: [in.idf](https://github.com/NREL/EnergyPlus/files/5195772/in.txt)

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by nealkruis**
Pull request overview
---------------------
 - Fixes #8262

Avoids reporting non-convex surfaces and related errors.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults

### CODE REVIEW

File: tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc [hunk 1]
```diff
@@ -290,6 +291,20 @@ bool EnergyPlusFixture::has_dfs_output(bool reset_stream)
...
+bool EnergyPlusFixture::match_err_stream(std::string const &expected_match, bool use_regex, bool reset_stream)
+{
+    auto const stream_str = this->err_stream->str();
+    bool match_found;
+    if (use_regex) {
+        match_found = std::regex_match(stream_str,std::regex(expected_match));
+    } else {
+        match_found = stream_str.find(expected_match) != std::string::npos;
+    }
+    EXPECT_TRUE(match_found);
```
Comment (nealkruis) [inline tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc]: This line may need to be removed depending on what we decide to do with `ENABLE_GTEST_DEBUG_MODE`.

File: tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc [hunk 2]
```diff
@@ -290,6 +291,20 @@ bool EnergyPlusFixture::has_dfs_output(bool reset_stream)
     return has_output;
 }
 
+bool EnergyPlusFixture::match_err_stream(std::string const &expected_match, bool use_regex, bool reset_stream)
```
Comment (nealkruis) [inline tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc:294]: @mbadams5 @nmerket let me know what you think about this function. If we like it, we can make similar functions for other streams.

File: src/EnergyPlus/PierceSurface.hh [hunk 3]
```diff
@@ -384,7 +384,7 @@ void PierceSurface(DataSurfaces::SurfaceData const &surface, // Surface
     //  Jan 2016: Initial release
 
     // Input checks
-    assert(std::abs(rayDir.mag_squared() - 1.0) < 4 * std::numeric_limits<Real64>::epsilon()); // Check unit vector
+    assert(std::abs(rayDir.mag_squared() - 1.0) < 6 * std::numeric_limits<Real64>::epsilon()); // Check unit vector (6x is rough estimate. Increase slightly as needed.)
```
Comment (nealkruis) [inline src/EnergyPlus/PierceSurface.hh:387]: This is a minor correction that was needed to debug the defect file.

**Comment by Myoldmopar** [comment]
Unit test failure needs to be fixed directly or this will be pushed.

**Comment by nealkruis** [comment]
@Myoldmopar: @mbadams5 and I talked about how the unit test failure is maybe a larger issue related to my question to @kbenne [here](https://github.com/NREL/EnergyPlus/commit/e75aac93c90c1e30e1d9c19dfb577c5074a8683c#r42440505).

Should I remove the internal EXPECT_TRUE from this unit test fixture function to allow it to pass for now, and come back to this question after the release?

**Comment by Myoldmopar** [comment]
That works for me. Let’s get this in.

**Comment by Myoldmopar** [comment]
Mac was reporting the unit test failure along with the other machines.  Once Mac reports success this will likely be fully successful and it will go in sometime this evening.

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
I guess that is fine.  @nealkruis can you confirm the SHD outputs being changed there?

**Comment by nealkruis** [comment]
Yes. The only thing that should be printed to the SHD file when using PixelCounting is instances where the subsurfaces is not bounded by the base surface. Other than that, it should be empty.

It does make me wonder if we should deprecate the SHD file in favor of including all the details in the ERR file. For another time, though...

**Comment by Myoldmopar** [comment]
Calling this one, thanks @nealkruis

**Comment by nealkruis** [comment]
@chriswmackey This has been merged for the v9.4 release at the end of September.

**Pull request merged**